### PR TITLE
Multiple IP per NIC - help files, PSObject, test and minor fixes

### PR DESCRIPTION
**Pull request by subashchandra31**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by Arpitaji** [comment]
@vidyadharijami is added to the review. #Closed

File: src/RecoveryServices/RecoveryServices.SiteRecovery.Test/ScenarioTests/A2A/AsrA2ATests.ps1 [hunk 1]
```diff
@@ -664,8 +664,12 @@ function Test-VMNicConfig {
     #Update VM Nic properties
     $pe = Get-AzRecoveryServicesAsrReplicationProtectedItem -ProtectionContainer $pc -Name  $vmName
     $nicId = $pe.NicDetailsList[0].NicId
+    $ipConfigName = $pe.NicDetailsList[0].IpConfigs[0].Name
     $recNicName = getRecoveryNicName
-    $nicConfig = New-AzRecoveryServicesAsrVMNicConfig -NicId $nicId -ReplicationProtectedItem $pe -RecoveryNicName $recNicName -RecoveryNicResourceGroupName $recRgName -ReuseExistingNic
+
+    $ipConfig = New-AzRecoveryServicesAsrVMNicIPConfig -IpConfigName $ipConfigName -RecoverySubnetName ""
+    $ipConfigs = @($ipConfig)
+    $nicConfig = New-AzRecoveryServicesAsrVMNicConfig -NicId $nicId -ReplicationProtectedItem $pe -RecoveryVMNetworkId $RecoveryAzureNetworkId -RecoveryNicName $recNicName -RecoveryNicResourceGroupName $recRgName -ReuseExistingNic -IPConfig $ipConfigs
```
Comment (punit1396) [inline src/RecoveryServices/RecoveryServices.SiteRecovery.Test/ScenarioTests/A2A/AsrA2ATests.ps1:672]: just comma separated should work in argument. no need for creating an array

File: src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs [hunk 2]
```diff
@@ -253,6 +253,28 @@ public override void ExecuteSiteRecoveryCmdlet()
                             vmNic.EnableAcceleratedNetworkingOnTfo ?? false;
                     }
 
+                    List<PSIPConfigInputDetails> ipConfigList = null;
+                    if (this.IPConfig == null || this.IPConfig.ToList().Count == 0)
+                    {
+                        ipConfigList = vmNic.IpConfigs?.Select(ip => ConvertToPSIPConfig(ip)).ToList() ?? null;
```
Comment (punit1396) [inline src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs:259]: ?.ToList()

File: src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs [hunk 3]
```diff
@@ -253,6 +253,28 @@ public override void ExecuteSiteRecoveryCmdlet()
...
+                    }
+                    else if (vmNic.IpConfigs != null)
+                    {
+                        ipConfigList = this.IPConfig.ToList();
+                        var inputIPConfigNames = this.IPConfig.Select(ip => ip.IPConfigName).ToList();
+
+                        foreach (IPConfigDetails ipConfig in vmNic.IpConfigs)
+                        {
+                            if (inputIPConfigNames.Contains(ipConfig.Name))
```
Comment (punit1396) [inline src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs:268]: case insensitive check

File: src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs [hunk 4]
```diff
@@ -253,6 +253,28 @@ public override void ExecuteSiteRecoveryCmdlet()
...
 
+                    List<PSIPConfigInputDetails> ipConfigList = null;
+                    if (this.IPConfig == null || this.IPConfig.ToList().Count == 0)
+                    {
+                        ipConfigList = vmNic.IpConfigs?.Select(ip => ConvertToPSIPConfig(ip)).ToList() ?? null;
+                    }
+                    else if (vmNic.IpConfigs != null)
+                    {
+                        ipConfigList = this.IPConfig.ToList();
+                        var inputIPConfigNames = this.IPConfig.Select(ip => ip.IPConfigName).ToList();
```
Comment (punit1396) [inline src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs:264]: This is only being used for Contains check. Contains should be present for IEnumerable. No need to call ToList()

File: src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs [hunk 5]
```diff
@@ -162,7 +162,7 @@ public class NewAzureRmAsrVmNicConfig : SiteRecoveryCmdletBase
             Mandatory = false,
             HelpMessage = "Specifies test failover/failover settings of NIC IP configs.")]
         [ValidateNotNull]
-        public IPConfigInputDetails[] IPConfig { get; set; }
+        public PSIPConfigInputDetails[] IPConfig { get; set; }
```
Comment (punit1396) [inline src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs:165]: Could you share the PS design review PR link? Wanted to go through that once.

Comment (subashchandra31) [reply src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs:165]: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/886

File: src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs [hunk 6]
```diff
@@ -363,5 +387,23 @@ private bool ValidateAndPopulateIPConfigs(ASRVMNicDetails vmNic)
 
             return true;
         }
+
+        PSIPConfigInputDetails ConvertToPSIPConfig(IPConfigDetails ipConfig)
```
Comment (punit1396) [inline src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs:391]: or you could write a mapper - check AutoMapper

**Review by punit1396 — CHANGES_REQUESTED** [review]

File: src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs (hunk 6 above)
Comment (subashchandra31) [reply src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs:391]: will stick with current approach as i see no performance benefit.

File: src/RecoveryServices/RecoveryServices.SiteRecovery.Test/ScenarioTests/A2A/AsrA2ATests.ps1 (hunk 1 above)
Comment (subashchandra31) [reply src/RecoveryServices/RecoveryServices.SiteRecovery.Test/ScenarioTests/A2A/AsrA2ATests.ps1:672]: leaving it for better readability as adding at cmdlet level in argument will make it more cumbersome

File: src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs (hunk 2 above)
Comment (subashchandra31) [reply src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs:259]: vmNic.IPConfigs is of type IList<> while IPConfigs in ASRVMNicConfig is of type List<>. So

File: src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs (hunk 3 above)
Comment (subashchandra31) [reply src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs]: done

File: src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs (hunk 4 above)
Comment (subashchandra31) [reply src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs]: done

File: src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs (hunk 2 above)
Comment (punit1396) [reply src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs:259]: Null check.

Comment (subashchandra31) [reply src/RecoveryServices/RecoveryServices.SiteRecovery/NetworkInterfaceCard/NewAzureRmAsrVmNicConfig.cs]: done

**Review by punit1396 — APPROVED** [review]

**Pull request merged**